File: oss/mcp-tooling/src/utils/sentry.ts
/**
 * Sentry Error Tracking Integration
 *
 * This module provides Sentry integration for the MCPOverflow frontend.
 * Install with: npm install @sentry/react
 */

import type { User } from '@/types/database';

// Type definitions for Sentry (will be available after npm install @sentry/react)
interface SentryOptions {
  dsn: string;
  environment: string;
  release?: string;
  tracesSampleRate: number;
  replaysSessionSampleRate: number;
  replaysOnErrorSampleRate: number;
  integrations?: any[];
  beforeSend?: (event: any, hint: any) => any;
}

interface SentryUser {
  id: string;
  email?: string;
  username?: string;
}

interface SentryScope {
  setUser(user: SentryUser | null): void;
  setTag(key: string, value: string): void;
  setContext(key: string, context: Record<string, any>): void;
}

// Placeholder types for when Sentry is not installed
let Sentry: any = null;
let isSentryAvailable = false;

// Try to import Sentry (will work after npm install @sentry/react)
try {
  Sentry = require('@sentry/react');
  isSentryAvailable = true;
} catch (error) {
  console.warn('Sentry not installed. Run: npm install @sentry/react');
}

/**
 * Initialize Sentry error tracking
 */
export function initSentry(): void {
  if (!isSentryAvailable || !Sentry) {
    console.warn('Sentry is not available. Error tracking disabled.');
    return;
  }

  const dsn = import.meta.env.VITE_SENTRY_DSN;
  if (!dsn) {
    console.warn('VITE_SENTRY_DSN not configured. Error tracking disabled.');
    return;
  }

  const environment = import.meta.env.VITE_ENVIRONMENT || import.meta.env.MODE || 'development';
  const release = import.meta.env.VITE_RELEASE || 'mcpoverflow@0.1.4';

  try {
    Sentry.init({
      dsn,
      environment,
      release,
      tracesSampleRate: environment === 'production' ? 0.1 : 1.0,
      replaysSessionSampleRate: environment === 'production' ? 0.1 : 0.0,
      replaysOnErrorSampleRate: 1.0,
      integrations: [
        Sentry.browserTracingIntegration({
          // Performance monitoring
          tracePropagationTargets: [
            'localhost',
            /^https:\/\/.*\.mcpoverflow\.(io|com|ai|dev)/,
          ],
        }),
        Sentry.replayIntegration({
          // Session replay for debugging
          maskAllText: true,
          blockAllMedia: true,
        }),
      ],
      beforeSend(event, hint) {
        // Filter out sensitive information
        if (event.request) {
          // Remove sensitive headers
          if (event.request.headers) {
            delete event.request.headers['authorization'];
            delete event.request.headers['cookie'];
            delete event.request.headers['x-api-key'];
          }

          // Remove sensitive query parameters
          if (event.request.query_string) {
            const sensitiveParams = ['token', 'key', 'secret', 'password'];
            sensitiveParams.forEach(param => {
              event.request.query_string = event.request.query_string?.replace(
                new RegExp(`${param}=[^&]*`, 'gi'),
                `${param}=***`
              );
            });
          }
        }

        // Don't send events in development unless explicitly enabled
        if (environment === 'development' && !import.meta.env.VITE_SENTRY_DEBUG) {
          return null;
        }

        return event;
      },
    });

    console.log(`Sentry initialized for environment: ${environment}`);
  } catch (error) {
    console.error('Failed to initialize Sentry:', error);
  }
}

/**
 * Set the current user context for Sentry
 */
export function setUser(user: User | null): void {
  if (!isSentryAvailable || !Sentry) return;

  try {
    if (user) {
      Sentry.setUser({
        id: user.id,
        email: user.email,
        username: user.email?.split('@')[0],
      });
    } else {
      Sentry.setUser(null);
    }
  } catch (error) {
    console.error('Failed to set Sentry user:', error);
  }
}

/**
 * Capture an exception manually
 */
export function captureException(error: Error, context?: Record<string, any>): void {
  if (!isSentryAvailable || !Sentry) {
    console.error('Sentry not available, error:', error, context);
    return;
  }

  try {
    if (context) {
      Sentry.withScope((scope: SentryScope) => {
        Object.entries(context).forEach(([key, value]) => {
          scope.setContext(key, value);
        });
        Sentry.captureException(error);
      });
    } else {
      Sentry.captureException(error);
    }
  } catch (err) {
    console.error('Failed to capture exception:', err);
  }
}

/**
 * Capture a message
 */
export function captureMessage(message: string, level: 'info' | 'warning' | 'error' = 'info'): void {
  if (!isSentryAvailable || !Sentry) {
    console.log(`[${level}] ${message}`);
    return;
  }

  try {
    Sentry.captureMessage(message, level);
  } catch (error) {
    console.error('Failed to capture message:', error);
  }
}

/**
 * Add a breadcrumb for tracking user actions
 */
export function addBreadcrumb(
  message: string,
  category: string = 'user-action',
  level: 'info' | 'warning' | 'error' = 'info',
  data?: Record<string, any>
): void {
  if (!isSentryAvailable || !Sentry) return;

  try {
    Sentry.addBreadcrumb({
      message,
      category,
      level,
      data,
      timestamp: Date.now() / 1000,
    });
  } catch (error) {
    console.error('Failed to add breadcrumb:', error);
  }
}

/**
 * Set a custom tag
 */
export function setTag(key: string, value: string): void {
  if (!isSentryAvailable || !Sentry) return;

  try {
    Sentry.setTag(key, value);
  } catch (error) {
    console.error('Failed to set tag:', error);
  }
}

/**
 * Set custom context data
 */
export function setContext(key: string, context: Record<string, any>): void {
  if (!isSentryAvailable || !Sentry) return;

  try {
    Sentry.setContext(key, context);
  } catch (error) {
    console.error('Failed to set context:', error);
  }
}

/**
 * Start a performance transaction
 */
export function startTransaction(name: string, operation: string): any {
  if (!isSentryAvailable || !Sentry) return null;

  try {
    return Sentry.startTransaction({ name, op: operation });
  } catch (error) {
    console.error('Failed to start transaction:', error);
    return null;
  }
}

/**
 * Finish a performance transaction
 */
export function finishTransaction(transaction: any): void {
  if (!transaction || !isSentryAvailable) return;

  try {
    transaction.finish();
  } catch (error) {
    console.error('Failed to finish transaction:', error);
  }
}

/**
 * Create an error boundary component
 * Use this to wrap your app or specific components
 */
export function createErrorBoundary(): any {
  if (!isSentryAvailable || !Sentry) {
    // Return a simple fallback error boundary
    return ({ children }: { children: React.ReactNode }) => children;
  }

  try {
    return Sentry.ErrorBoundary;
  } catch (error) {
    console.error('Failed to create error boundary:', error);
    return ({ children }: { children: React.ReactNode }) => children;
  }
}

/**
 * Track API errors specifically
 */
export function captureAPIError(
  error: Error,
  endpoint: string,
  method: string,
  statusCode?: number
): void {
  captureException(error, {
    api: {
      endpoint,
      method,
      statusCode,
    },
  });
}

/**
 * Track navigation/routing errors
 */
export function captureNavigationError(error: Error, path: string): void {
  captureException(error, {
    navigation: {
      path,
      referrer: document.referrer,
    },
  });
}

/**
 * Track authentication errors
 */
export function captureAuthError(error: Error, action: string): void {
  captureException(error, {
    auth: {
      action,
      timestamp: new Date().toISOString(),
    },
  });
}

// Export whether Sentry is available
export const sentryAvailable = isSentryAvailable;

// Export Sentry for advanced usage
export { Sentry };

File: oss/mcp-tooling/services/api-service/internal/monitoring/sentry.go
package monitoring

import (
	"fmt"
	"log"
	"time"

	"github.com/getsentry/sentry-go"
	sentrygin "github.com/getsentry/sentry-go/gin"
	"github.com/gin-gonic/gin"
)

// SentryConfig holds Sentry configuration
type SentryConfig struct {
	DSN              string
	Environment      string
	Release          string
	TracesSampleRate float64
	Debug            bool
	AttachStacktrace bool
}

// InitSentry initializes Sentry error tracking
func InitSentry(config SentryConfig) error {
	if config.DSN == "" {
		log.Println("Sentry DSN not configured, error tracking disabled")
		return nil
	}

	err := sentry.Init(sentry.ClientOptions{
		Dsn:              config.DSN,
		Environment:      config.Environment,
		Release:          config.Release,
		TracesSampleRate: config.TracesSampleRate,
		Debug:            config.Debug,
		AttachStacktrace: config.AttachStacktrace,
		BeforeSend: func(event *sentry.Event, hint *sentry.EventHint) *sentry.Event {
			// Filter out sensitive information before sending to Sentry
			if event.Request != nil {
				// Remove sensitive headers
				if event.Request.Headers != nil {
					delete(event.Request.Headers, "Authorization")
					delete(event.Request.Headers, "Cookie")
					delete(event.Request.Headers, "X-API-Key")
				}

				// Remove sensitive query parameters
				if event.Request.QueryString != "" {
					// You can add custom filtering here
				}
			}

			return event
		},
	})

	if err != nil {
		return fmt.Errorf("failed to initialize Sentry: %w", err)
	}

	log.Printf("Sentry initialized for environment: %s", config.Environment)
	return nil
}

// SentryMiddleware returns a Gin middleware that integrates with Sentry
func SentryMiddleware() gin.HandlerFunc {
	return sentrygin.New(sentrygin.Options{
		Repanic:         true,
		WaitForDelivery: false,
		Timeout:         2 * time.Second,
	})
}

// CaptureError captures an error and sends it to Sentry
func CaptureError(err error) {
	if err != nil {
		sentry.CaptureException(err)
	}
}

// CaptureErrorWithContext captures an error with additional context
func CaptureErrorWithContext(err error, context map[string]interface{}) {
	if err != nil {
		sentry.WithScope(func(scope *sentry.Scope) {
			for key, value := range context {
				scope.SetExtra(key, value)
			}
			sentry.CaptureException(err)
		})
	}
}

// CaptureMessage captures a message and sends it to Sentry
func CaptureMessage(message string, level sentry.Level) {
	sentry.WithScope(func(scope *sentry.Scope) {
		scope.SetLevel(level)
		sentry.CaptureMessage(message)
	})
}

// SetUser sets the current user context for Sentry
func SetUser(c *gin.Context, userID string, email string, username string) {
	if hub := sentrygin.GetHubFromContext(c); hub != nil {
		hub.ConfigureScope(func(scope *sentry.Scope) {
			scope.SetUser(sentry.User{
				ID:       userID,
				Email:    email,
				Username: username,
			})
		})
	}
}

// SetTag sets a custom tag for the current scope
func SetTag(c *gin.Context, key string, value string) {
	if hub := sentrygin.GetHubFromContext(c); hub != nil {
		hub.ConfigureScope(func(scope *sentry.Scope) {
			scope.SetTag(key, value)
		})
	}
}

// SetContext sets custom context data
func SetContext(c *gin.Context, key string, value map[string]interface{}) {
	if hub := sentrygin.GetHubFromContext(c); hub != nil {
		hub.ConfigureScope(func(scope *sentry.Scope) {
			scope.SetContext(key, value)
		})
	}
}

// AddBreadcrumb adds a breadcrumb to track user actions
func AddBreadcrumb(c *gin.Context, message string, category string, level sentry.Level) {
	if hub := sentrygin.GetHubFromContext(c); hub != nil {
		hub.AddBreadcrumb(&sentry.Breadcrumb{
			Message:   message,
			Category:  category,
			Level:     level,
			Timestamp: time.Now(),
		}, nil)
	}
}

// CaptureGinError captures a Gin context error
func CaptureGinError(c *gin.Context, err error) {
	if err == nil {
		return
	}

	if hub := sentrygin.GetHubFromContext(c); hub != nil {
		hub.WithScope(func(scope *sentry.Scope) {
			// Add request context
			scope.SetContext("request", map[string]interface{}{
				"method":      c.Request.Method,
				"url":         c.Request.URL.String(),
				"user_agent":  c.Request.UserAgent(),
				"remote_addr": c.ClientIP(),
			})

			// Add user context if available
			if userID, exists := c.Get("user_id"); exists {
				scope.SetUser(sentry.User{
					ID: fmt.Sprintf("%v", userID),
				})
			}

			hub.CaptureException(err)
		})
	}
}

// Flush waits for all events to be sent to Sentry
// Call this before shutting down the application
func Flush(timeout time.Duration) bool {
	return sentry.Flush(timeout)
}

// Performance monitoring helpers

// StartTransaction starts a new transaction for performance monitoring
func StartTransaction(c *gin.Context, name string, operation string) *sentry.Span {
	if hub := sentrygin.GetHubFromContext(c); hub != nil {
		transaction := sentry.StartTransaction(c,
			name,
			sentry.WithOpName(operation),
		)
		c.Set("sentry_transaction", transaction)
		return transaction
	}
	return nil
}

// StartSpan starts a child span for the current transaction
func StartSpan(c *gin.Context, operation string, description string) *sentry.Span {
	transaction, exists := c.Get("sentry_transaction")
	if !exists {
		return nil
	}

	if tx, ok := transaction.(*sentry.Span); ok {
		span := tx.StartChild(operation)
		span.Description = description
		return span
	}

	return nil
}

// FinishSpan finishes a span
func FinishSpan(span *sentry.Span) {
	if span != nil {
		span.Finish()
	}
}

// RecordError records an error in the current span
func RecordError(span *sentry.Span, err error) {
	if span != nil && err != nil {
		span.Status = sentry.SpanStatusInternalError
		span.SetData("error", err.Error())
	}
}
